docs/codestyle/javascript.rst

```diff
@@ -0,0 +1,1105 @@
+
+**Bad**:
+
+.. code-block:: django
+
+    <div class="js-component-1 js-component-2"
+        data-something="false" {# for component 2 #}
+        data-value="for component 1"
+        data-value="for component 2"> {# aw maaan #}
+        Sad panda :(
+    </div>
+
+Imagine in this case component 1 functionality is significantly affected by an option that is meant for component 2.
+Also if they share the same option property name, such as value - sad panda.
+
+**Good**:
+
+.. code-block:: django
+
+    <div class="js-component-first js-component-second"
+        data-component-first='{
+            "value": "for component 1"
+        }'
+        data-component-second='{
+            "value": "for component 2",
+            "something": false
+        }'>
+        Happy panda!
+    </div>
+
+Passing the data to the components is also very straightforward. This way you have the same initialisation method
+for all existing instances of the widget even if they have different options.
+
+.. code-block:: javascript
+
+    var componentElements = $('.js-component-2');
+    var defaults = {
+        x: 0,
+        y: 0,
+        something: true
+    };
+    componentElements.each(function () {
+        var componentElement = $(this);
+        var options = $.extend({}, defaults, componentElement.data('component-second'));
+        new ComponentSecond(componentElement, options);
+    }):
+
+
+Magic numbers
+-------------
+
+- Avoid magic numbers. Try to parametrise or use constants.
+
+.. code-block:: javascript
+
+    // bad
+    setTimeout(function () {
+        if (failed && count < 5) {
+            count++;
+            return;
+        }
+        // or do stuff
+    }, 3000);
+
+    // better
+    var POLLING_TIMEOUT = 3000;
+    var MAX_FAILURES_COUNT = 5;
+
+    setTimeout(function () {
+        if (failed && count < MAX_FAILURES_COUNT) {
+            count++;
+            return;
+        }
+        // or do stuff
+    }, POLLING_TIMEOUT);
+
+
+.. code-block:: javascript
+
+    switch (e.keyCode) {
+        case keyCodes.ENTER:
+        case keyCodes.SPACE:
+            x();
+            break;
+        case keyCodes.TAB
+        case keyCodes.ESCAPE:
+            y();
+            break;
+        default:
+            z();
+    }
+
+
+ECMAScript 5
+============
+
+Use where appropriate. Use array methods for working with arrays, but don't use them when working with array-like
+objects such as jQuery collections. For them use ``$.fn.each`` instead.
+
+Prefer ``Array#forEach`` over ``for () {}`` loop.
+
+.. code-block:: javascript
+
+    var fighters = [
+        {
+            name: 'Jonny Cage',
+            dead: true
+        },
+        {
+            name: 'Kung Lao',
+            dead: true
+        },
+        {
+            name: 'Raiden',
+            dead: false
+        }
+    ];
+
+    // bad
+    var i;
+    var l = fighters.length;
+
+    for (; i < l; i++) {
```

Comment:
this is also quite questionable performance-wise: https://jsperf.com/es5-object-looping